### ISSUE TITLE
what are the default methods?

### ISSUE THREAD

**Issue by kjx** [gracelang/language#49]
This issue is to resolve what the "default methods" actually are.   We expect every object that is an instance of a class will get these default methods somehow.  This issue is a **different question** to whether every object gets these default methods, or just most of them #38, #39

### spec
1. == operator: the implementation inherited from graceObject answers true if and only if the argument is the same object as the receiver.
1. !=, which answers the negation of ==
1. hash, which must be compatible with ==. This means that two objects that are == must have the same hash
1. asString, which answers a string representation of the object tailored for the end user, and
1. asDebugString, which answers a string representation of the object tailored for the programmer; the implementation inherited from graceObject answers asString.

Also:

> The only requests understood by done are asString and asDebugString. (In particular, done does not have an equality method).

> Immutable objects (objects that have no var fields and that capture no mutable objects) that the programmer wishes to treat as a “value objects” should override == so that it implements Leibniz equality.

### buildinTypesInGrace
   == (other: Object) -> Boolean
   != (other: Object) -> Boolean
   /= (other: Object) -> Boolean
   $\neq$ (other: Object) -> Boolean
   // the inverse of ==.   All three variants have the same meaning.
   hash -> Number
   match (other: Object) -> SucccessfulMatch | FailedMatch
   asString -> String
   asDebugString -> String
   :: (other:Object) -> Binding

### amg 
        addmethod2(GraceDefaultObject, "asString", &Object_asString);
        addmethod2(GraceDefaultObject, "++", &Object_concat);
        addmethod2(GraceDefaultObject, "==", &UserObj_Equals);
        addmethod2(GraceDefaultObject, "!=", &Object_NotEquals);
        addmethod2(GraceDefaultObject, "≠", &Object_NotEquals);
        addmethod2(GraceDefaultObject, "asDebugString", &Object_asDebugString);
        addmethod2(GraceDefaultObject, "basicAsString", &Object_asString);
        addmethod2(GraceDefaultObject, "::", &Object_bind);

### Kernan 
                AddMethod("asString", null);
                AddMethod("==", null);
                AddMethod("!=", null);

**Comment by kjx** [comment]
notably, there are a range of possibilities  (this text also in #38)
 1. objects don't have any default methods: trying to call them leads to a dynamic type error
 1. objects have default methods but the _object_ is obviously **incomplete** (a direct instance of a trait or an abstract class) 
 1. objects have default methods but we know they can obviously never work (they are abstract or trait conflicts or ... #59)
 1. objects have default methods that non-obviously never work (a template method than calls another method that is abstract or trait conflict)
 1. objects have default methods but the object is uncooked #56 
 1. objects have dealt methods that just break. 

One consequence of this is that any systems code, particularly in debugging etc, can **never** really depend on default methods: this kind of defensive coding is always required: https://github.com/gracelang/language/issues/39#issuecomment-183551871

**Comment by kjx** [comment]
trying to carry on the _semantic_ discussion (i.e. how default method work with traits) in #38.
Let's just talk about the list here.

**Comment by kjx** [comment]
Mark Miller said that JS (and his E) builds in identity hash tables but NOT hash codes --- so hashcode doesn't have to be a default method if we had some magic like that...

**Comment by kjx** [comment]
What can you delete a default method?  argh!

**Comment by apblack** [comment]
_Minigrace_ does not build in `identityHash`, and I don't see any easy way of rth programmer to provide it.  It seems that it would be necessary for the object to cons-up its own uid using a counter, and then hash it, or make it probabilistic and use a pseudo-random number..

Here is an example of the benefit of having `done` understand `asString`.  The scenario is that a method returns `done` because some code was added after the expression that calculates the answer — an easy mistake for anyone to make.

```
No method 'dtype:=' ; available methods are:
  asDebugString  asString
NoSuchMethod at line 26 of ast: no method 'dtype:=' in done.
  called from done.dtype:= (defined nowhere) at ast:26
  called from obj6783.new.map()ancestors (defined at ast:1674 in obj6783.new created at ast:1630) at ast:2037
...
```

Before @kjx suggests it, yes, it would be possible in theory to check for the existence of `asString` just before requesting it, every time it is requested.  This involves reflection, and is likely to be slow.  It is also infeasible in practice: people just won't do it.  More importantly, the whole point of OO programming is that one does _not_ design interfaces so that programmers have to make such tests — you make the request and it _just works_.  Grace should be graceful, and Graceful programs should be paradigmatic examples of OOP.

**Comment by kjx** [comment]
> Minigrace does not build in identityHash, and I don't see any easy way for the programmer to provide it. 

yep - are you suggesting it should?   I fear this is related to what the model of equality and identity is (or should be) which is something else we never really settled on. #91 

> The scenario is that a method returns done because some code was added after the expression that calculates the answer — an easy mistake for anyone to make.

Right. it's a _mistake_ - and one that would easily be caught by a static type checker if the return type is declared, and if that type excludes `done`.  If I write `-> type { asString }`  do I really want that to permit returning `done`?     I don't see the conceptual rationale for having `asString` on `done` when we don't have `==` and friends: one could always write something like `"done" == x.asString` to get the effect of `==` on done.  I guess I see `==`  as the most basic default method (and these days don't even like that).

> yes, it would be possible in theory to check for the existence of asString just before requesting it, every time it is requested. This involves reflection

I'm *absolutely not* suggesting this as general programming practice.

I do think it is necessary, e.g. if you're writing debuggers etc, because in incorrect programs you cannot rely on `asString` returning anything, or on it actually returning a String.   

If we believe that programmers using `done` as anything other than a placeholder for a non-result , then trying to do any of that should be an error.     Rather than checking on `done`, people could do type tests and should generally rule out `done`, or better still, should pass real objects around rather than `done`. 

> More importantly, the whole point of OO programming is that one does not design interfaces so that programmers have to make such tests — you make the request and it just works. 

absolutely. But `done` getting out there surely a case where something has gone wrong: and I want to catch that asap - or at least not make it more difficult to catch.

> Grace should be graceful, and Graceful programs should be paradigmatic examples of OOP

yes and yes.  The difficult question is: what is more graceful or more OO?  
In this case, the issues are: how should type tests work with things like `done`, and how should objects get the default methods (which can make other bits of the language ungraceful).  

There are conceptual differences between _methods on *all* objects_ and _methods on *most* objects_ 
This is one of the differences between Smalltalk (all) and Self (most). 

If we go for a _most objects_ design then 
 - done has no methods
 - we admit there are some other (rare, atypical) objects that may have no methods, or may not have all the default methods 
 - for example instances of raw traits can be OK even if missing default methods, or if those default methods are `required` (was abstract) 
 - we can juggle the relationship between `class` `trait` and `object` to bring in default methods, or not. 
- If we think `print` is at least partly for debugging, we could special case print to print `done` #93, and to tolerate other errors .e.g when `asString` is broken.

**Comment by kjx** [comment]
See also my attempt at making a table of the core types #96

**Comment by KimBruce** [comment]
I'd prefer to see all objects (including done) and classes provide the same methods, and to have them include whatever equality and inequality methods we want as well asString and asDebugString.

Yes, some people will do strange things with them, but its not worth overly complicating the language for such limited benefit.

**Comment by KimBruce** [comment]
I meant to add that I consider traits to be restricted (deficient?) objects.  They don't have defs, vars, initialization code, or default methods.

**Comment by kjx** [comment]
I like the phrase _deficient objects_, and that traits don't get stuff including default methods.   That means we can avoid special-casing the default methods in the inheritance algorithm (I hope!). 

The only question is which way around the _equivalence_ between shortcut forms and object constructors work:

| write |  Option A | Option B | 
|:-----|:-----|:-----|
| `class foo { }`   | `object { inherits graceObject }` | `object { }` | 
| `trait foo { }`   | `object { }` | `object { inherits nothing }` | 

I prefer option A, but I think you'll prefer option B. 

(where `nothing` would ideally be `done` but that's another issue...)

**Comment by KimBruce** [comment]
I do definitely prefer your option b.  This is because I want to go smoothly from objects to classes in the first week of classes and I want to just have to explain that classes are simply object generators.  I do not want to have to talk about the default methods yet.  (I probably won’t even talk about traits in the first course.)

Also, because traits are deficient objects, defining one as object{} seems weird to me.  Object{} should have the methods.

Kim

**Comment by kjx** [comment]
This _draft essay_ talks about the equals and not-equals methods.

[essay.pdf](https://github.com/gracelang/language/files/204157/essay.pdf)

code from the end of the essay: 

```
type Object = interface {
  == (other : Object) -> Boolean
  != (other : Object) -> Boolean
  hash -> Boolean
  ...
}

trait graceObject {
  method refEqualsSelf(other : Object)
    is primitive, confidential { } 
  method == (other : Object) {refEqualsSelf(other)}
  method != (other : Object) {! (self == other)}
  ...
}

trait graceValue { 
  use graceObject
  method publicFieldsEquals(other : Object) 
    is primitive, confidential { } 
  method == (other : Object) {publicFieldsEquals(other)}
  ...
}


type Reflection = interface { 
  reflect(reflectee : Object) -> ObjectMirror
  ...
}

type ObjectMirror = Object & interface {
  ...
}
```

**Comment by kjx** [comment]
After Michael pointed it out, I quite like the the Python solution: everything has equals, but effectively only immutable objects have hash.  So you couldn't use mutable collections as keys in collections...

**Comment by kjx** [comment]
Adopted

buildinTypesInGrace

== (other: Object) -> Boolean
!= (other: Object) -> Boolean
/= (other: Object) -> Boolean
$\neq$ (other: Object) -> Boolean
// the inverse of ==. All three variants have the same meaning.
hash -> Number
asString -> String
asDebugString -> String
:: (other:Object) -> Binding

**Comment by apblack** [comment]
The Python solution, then, would exclude `hash` from the methods in object.  

But that's not necessary, if the default `hash` is identity hash and the default `==` is identity, then `hash` and `==` correspond, and any object can indeed be used as a key.

Of course,if one were to override `==`, then one would have to make sure that one also overrode `hash`, but that's always true.

**Comment by kjx** [comment]
> Of course, if one were to override ==, then one would have to make sure that one also overrode hash, but that's always true

the point of the Python design is a) that hash must be consistent with === and b) hash must be stable.
If == isn't stable - if it depends on mutable state - then Python does **not** include a hash method, ad that acts as a signal that such an object cannot be used as a Key in a collection. 

There are two questions here 
 - should mutable collections (in particular) have a 'hash' method?    (applies to other mutable objects too)
 - what should that 'hash' method do?

solutions include
 - Java/Smalltak, everything has both == and hash, == on mutable collections is "equals now", hash is not stable.  Can use mutcols as keys provided no-one mutates them
 - EGAL - == and hash on mutable collection is identity.  I think that's really confusing! 
 - Python - mutable collections/objects don't offer hash. == is equals now. must convert to immutable or something before using as a key in a collection    (hence graceObject can't have a public hash method at least)

**Comment by apblack** [comment]
To summarize what @kjx writes [above](https://github.com/gracelang/language/issues/49#issuecomment-210669441):

 * The JavaSmalltalk solution required that programmers know what they are doing — if they use a mutable object as a key, they they better not mutate it!

 * The Python solution is more robust — you will get an error as soon as you try to use a mutable object as a key.  The cost is efficiency: you have to convert a list into a tuple before you can use it as a key.

Phrased this way, given our unconcern with efficiency, and our target audience, it seems that the Python approach is the one that we should follow.  But note: 
 
* it works only with _hashed_ collections.  Collections implemented by trees can still use mutable keys, 
and they will break if the values are mutated in a way that changes the sort order;

* it obliges us to provide a more complete collection library, including, in particular, immutable sets and dictionaries;

* user-defined objects can still implement `hash`.

**Comment by apblack** [comment]
I've been reading _the left-hand of Equality_ essay.  It's very good, by the way.  It also has implications for this issue: the default methods.

By the autognostic principle, `graceObject` should *not* understand = and ≠.  Instead, it should support a *confidential* method `isMe(other)` that answers `true` if and only if `other` is identical to the target — which must also be the requester, precisely because `method isMe(other) is confidential`.

We can also have a `trait identity` that defines `=` and `≠` in terms of `isMe(_) *if* an object chooses to use that trait.

And we should take `identical` and `different` out of the `prelude`.

**Comment by kjx** [comment]
Well I'd be happy with that, actually. No doubt we should talk a bit more

**Comment by kjx** [comment]
On equals-now, equals-always, and hash:

If we were happy with two equals operators then we could have both equals-now and equals-always - but (for good reasons) I'd like to get by with just one.    That's possible because if the core specification is equals-now, equals-always satisfies that specification.  The question is how we signal which contract our single equals implementation meets.   Python more or less uses the presence or absence of a "hash" method to signal that. This is an issues with structural types: we only have method names to signal semantics, so really we don't have a good option.  If we had *nominal* types, then we could use e.g. different Java-style interfaces to signal the semantics, but we don't' have that option.

**Comment by apblack** [comment]
> The question is how we signal which contract our single equals implementation meets

The way that we have adopted to communicate the meaning of methods in an interface is comments in th type.  That's not fully adequate, of course, but it's the best we have.  Is there a reason to worry about the specification of `==` more than some other method?

That was a question, and not rhetoric.  If the answer is "yes", then perhaps we _should_ have more than one symbol for equality.  In other words, if we believe that programmer really need to be concerned with the distinction between equal-now and equal-always, then we should use different operators.

**Comment by kjx** [comment]
_Is there a reason to worry about the specification of == more than some other method?_

If there is, it's because collections are really important: the methods are ==, !=, hash, compareTo (or <=> or however we write it...) 

_if we believe that programmer really need to be concerned with the distinction between equal-now and equal-always, then we should use different operators_

Yes I can see the rationale for that, it's in the paper. But I'd still like just one.  There is another rationale for just one which might also be in the paper (not sure): that the same object(s) having different == _semantics_ is also confusing.  That's what I like about "equal now by default, some way to signal equal-always"

_ [we] communicate the meaning of methods in an interface is comments in th type. .. but it's the best we have_  

Syntactically we also have annotations.  We could make (some?) annotations take part in type checking. Particularly in this case, where the issue is really (deep) (im)mutability, there may well be other methods were frameworks will want to know something isn't going to change. Something like this: 

```
type Eq = interface { 
   == (_) -> Boolean 
   hash -> Number 
} 

type Key = { 
   == (_) -> Boolean is pure
   hash -> Number is pure 
} 
``` 

Of course, that doesn't help on the "dynamic" side of things, presumably the default purity must be something like Unknown that will match anyway: and a contract to catch that violation while possible would be rather... slow.    grrr.

**Comment by apblack** [comment]
> that the same object(s) having different == semantics is also confusing

Sorry, bit I've no idea what that means.

I'm also not sure what you mean to imply with your `is pure` annotation.  "Effect-free" is what it says to me.   But that would apply to all `==` and `hash` methods, I would hope.

I'm happy to talk about putting immutability into types.

Back to the subject of equality, there is an object that the essay doesn't consider: `antiPerverse`.
It's never equal to anything.  And it exists, per the IEEE: "NaN".  So we get 

```
def NaN = 1/0
assert(NaN ≠ NaN) description "We break the IEEE floating point standard"
assert(identical(NaN, NaN)) description "if that's not true, all is lost!"
```

The idea of left-handed equality avoid this paradox, because we get rid of `identical` as a method on the prelude, and make sure that NaN does not implement `isMe`.

    assert{ NaN.isMe(NaN) } shouldRaise (NoSuchMethod)

**Comment by kjx** [comment]
_that the same object(s) having different == semantics is also confusing_

once you have more than one equals operator, the same object can compare differently depending on which operator you use..

_I'm also not sure what you mean to imply with your is pure annotation._

I probably mean "does not depend on mutable state" - ruling out read effects as well as write effects. 
Quite easy to do dynamically & inefficiently: doing it statically and efficiently is tricker.

**Comment by apblack** [comment]
> once you have more than one equals operator, the same object can compare differently depending on which operator you use.

Of course.  If the two operators _always_ give the same answer, what's the point to having two of them?

**Comment by KimBruce** [comment]
I've just run into this issue while converting some Grace code to new syntax (though the problem has nothing to do with the syntax.  Take the following example:

```
def o = object {
    method asString {"Dummy"}
    method ==(other:Object) {self.isMe(other)}
    method !=(other: Object) {!(self == other)}
}

print (o == o)
```
Minigrace (annoyingly) insisted that I should include the definition of == as written above.

I think everything should have a primitive == that supports identity.  Of course the programmer can override it (at her own peril).  I can't imagine an instance where I wouldn't want to have both == and !=.

One could ease the pain very slightly by including an equality trait in the standard prelude, leading to

```
def o = object {
    use equality
    method asString {"Dummy"}
}

trait equality {
    method ==(other:Object) {self.isMe(other)}
    method !=(other: Object) {!(self == other)}
}
```
However, I still believe everything should have == and !=

**Comment by KimBruce** [comment]
I'm still struggling with equality (though I did finally read through James' screed on equality -- very nice!!).  We seem to have a top type Object that includes == and != (see section 10.1.2 of the spec).  The spec in 6.5.4 also states "All objects implement a number of default methods by inheriting from the graceObject trait."  

yet if I define a class, it does not automatically have those defined as methods (at least not in minigrace), though isMe is there to be used in defining my own == (and != is defined as the opposite of ==).

If I try using == on a new class (without defining ==) then I get the following error message:
NoSuchMethod on line 71 of my file: no method '==' on myClass.

I think Object should reflect the operations pre-defined on an empty class:  class c { }

Thus if we don't think everything should have ==, then it should be removed from Object.

To make life easier for programmers, the standard prelude should include the following trait:
```
trait equalTrait [[T]] {
     method ==(other:T) -> Boolean {self.isMe(other)}
     method != (other:T) -> Boolean {!(self == other)}
}
```

Is that a reasonable compromise?

**Comment by apblack** [comment]
You are right, I think.

= and ≠ should be removed from Object.  I believe that it's only because of my oversight that they are there.

That trait should be in the standard prelude, too.  But traits don't yet work in the C backend.  So, for compatibility, I have been adding the method `==`, to those objects that need it, by hand.  

For this reason, it's been convenient to leave ≠ in `graceObject`; that way one has to add only one method, and one is pretty certain that `≠` will be the inverse of == (unless the programmer deliberately defines it as something else!)
